File: codility/10-Flags.js
function solution(A) {
  // Map all peak indices so we could later compare distances
  const peaks = [];
  for (let i = 1; i < A.length - 1; i++) {
    const isPeak = A[i] > A[i - 1] && A[i] > A[i + 1]; // It is a peak if left and right points are lower
    if (isPeak) peaks.push(i);
  }

  // If there are fewer than 2 peaks, distances do not matter
  if (peaks.length < 3) return peaks.length;

  // Max possible number of flags is a root of the distance between first and the last peak
  const firstPeakIdx = peaks[0];
  const lastPeakIdx = peaks[peaks.length - 1];
  const maxFlags = Math.floor(Math.sqrt(lastPeakIdx - firstPeakIdx) + 1);

  // Iterate reverse from max possible flags to minimum (2) and check if all peaks match requirements
  // Minimum number of flags to try is 3: 2 flags can be planted an oany 2 peaks (2 is the minimum possible distance between peaks)
  for (let i = maxFlags; i > 2; --i) {
    // We need i number of flags in total: Check if sum of planted flags and remaining peaks to check is enough to match requirements (i)
    let plantedFlags = 1; // Always plant one (last) peak
    let lastPlanted = peaks.length - 1;
    let testedPeak = peaks.length - 2;

    const arePeaksInvalid = () => plantedFlags + testedPeak + 1 < i;
    while (!arePeaksInvalid()) {
      // Get distance from last planted flag
      const distance = peaks[lastPlanted] - peaks[testedPeak];
      // Peak is valid if distance is larger or equal to currently tested scenario (i)
      const isFarEnough = distance >= i;
      // If valid, plant flag
      if (isFarEnough) {
        plantedFlags++;
        lastPlanted = testedPeak;
        if (plantedFlags === i) return i; // Return early if max possible flags already planted
      }
      // Test next peak (reverse direction)
      testedPeak--;
    }

    // If planted all i required flags, return
    if (plantedFlags === i) return i;
  }
}

const arr = [1, 5, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2];
// const arr = [0, 0, 0, 0, 0, 1, 0, 1, 0, 1]
const s = solution(arr);

File: codility/py/10-flags.py
import math


def fn(mountain):
    if len(mountain) < 3:
        return 0

    # Record peak indexes
    peaks = []
    for i in range(1, len(mountain) - 1):
        if mountain[i - 1] < mountain[i] > mountain[i + 1]:
            peaks.append(i)

    if len(peaks) < 2:
        return len(peaks)

    # Determine max possible number of flags
    # k = total_flags and flag_distance
    k = math.ceil((peaks[-1] - peaks[0]) ** 0.5)

    # Required total distance is pow2 of k reduces for one step ok
    # It is because first flag can be planted on first peak and steps are counted starting from there
    required_distance = k * (k - 1)

    # Due to math.ceil, this will happen when distance between first and last peak is bigger than a perfect squar
    # Reduce for 1
    if required_distance > peaks[-1] - peaks[0]:
        k -= 1

    """
    Lower possible k while checking if flags can be planted according to the task rules starting from second peak
    """
    while k > 2:
        planted = 1  # Always plant first flag on first peak
        last_planted = peaks[0]  # Keep track of last planted idx
        for i in range(1, len(peaks)):
            # Only plant flag if far enough from last planted and enough room for following flags
            min_pos = last_planted + k
            max_pos = peaks[-1] - k * (k - planted - 1)
            if min_pos <= peaks[i] <= max_pos:
                planted += 1
                last_planted = peaks[i]
                if planted == k:
                    return k

    return k


assert fn([1, 5, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2]) == 3
